src/chain/link.rs

```diff
@@ -0,0 +1,238 @@
+use std::ops::Range;
+
+use hal::command::CommandBuffer;
+use hal::memory::Dependencies;
+use hal::pso::PipelineStage;
+use hal::queue::{QueueFamilyId, Supports, Transfer};
+
+use resource::{Access, Layout, Resource};
+use queue::QueueId;
+
+#[derive(Clone, Copy, Debug)]
+pub(super) struct Acquire;
+#[derive(Clone, Copy, Debug)]
+pub(super) struct Release;
+
+pub(super) trait Semantics {
+    fn src_dst(this: QueueFamilyId, other: QueueFamilyId) -> (QueueFamilyId, QueueFamilyId);
+}
+
+impl Semantics for Acquire {
+    fn src_dst(this: QueueFamilyId, other: QueueFamilyId) -> (QueueFamilyId, QueueFamilyId) {
+        (other, this)
+    }
+}
+
+impl Semantics for Release {
+    fn src_dst(this: QueueFamilyId, other: QueueFamilyId) -> (QueueFamilyId, QueueFamilyId) {
+        (this, other)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(super) enum LinkSync<A, L, S, M> {
+    /// No transition required.
+    None(M),
+
+    /// Pipeline barrier.
+    Barrier {
+        access: Range<A>,
+        layout: Range<L>,
+        stages: Range<PipelineStage>,
+    },
+
+    /// Signal / Wait for semaphore.
+    Semaphore { semaphore: S },
```

Comment:
it does appear that the enum could also be represented by a struct with optional fields:
```rust
struct LinkSys {
  barrier: Option<Barrier>,
  semaphore: Option<S>,
  transfer: Option<QueueId>,
}
```

Reply:
Except that having some queue without barrier and semaphore is nonsence.